mypy/types.py

```diff
@@ -1486,112 +1486,112 @@ def keywords_str(self, a: Iterable[Tuple[str, Type]]) -> str:
         ])
 
 
-# These constants define the method used by TypeQuery to combine multiple
-# query results, e.g. for tuple types. The strategy is not used for empty
-# result lists; in that case the default value takes precedence.
-ANY_TYPE_STRATEGY = 0   # Return True if any of the results are True.
-ALL_TYPES_STRATEGY = 1  # Return True if all of the results are True.
+# Combination strategies for boolean type queries
+def ANY_TYPE_STRATEGY(current: bool, accumulated: bool) -> bool:
+    """True if any type's result is True"""
+    if accumulated:
+        raise ShortCircuitQuery()
+    return current
 
 
-class TypeQuery(TypeVisitor[bool]):
-    """Visitor for performing simple boolean queries of types.
+def ALL_TYPES_STRATEGY(current: bool, accumulated: bool) -> bool:
+    """True if all types' results are True"""
+    if not accumulated:
+        raise ShortCircuitQuery()
+    return current
 
-    This class allows defining the default value for leafs to simplify the
-    implementation of many queries.
-    """
 
-    default = False  # Default result
-    strategy = 0     # Strategy for combining multiple values (ANY_TYPE_STRATEGY or ALL_TYPES_...).
+class ShortCircuitQuery(Exception):
+    pass
 
-    def __init__(self, default: bool, strategy: int) -> None:
-        """Construct a query visitor.
 
-        Use the given default result and strategy for combining
-        multiple results. The strategy must be either
-        ANY_TYPE_STRATEGY or ALL_TYPES_STRATEGY.
-        """
+class TypeQuery(Generic[T], TypeVisitor[T]):
+    """Visitor for performing queries of types.
+
+    default is used as the query result unless a method for that type is
+    overridden.
+
+    strategy is used to combine a partial result with a result for a particular
+    type in a series of types.
+
+    Common use cases involve a boolean query using ANY_TYPE_STRATEGY and a
+    default of False or ALL_TYPES_STRATEGY and a default of True.
+    """
```

Comment:
Maybe get rid of `TypeQuery.default`, and instead store it inside the strategy? In that case, all methods that currently `return self.default` would call `self.query_types([])` instead. This would make things slightly more streamlined, as everything will be routed through `strategy` (in particular, a strategy would even be able to change the default for query result in case the list is empty).